coast/tidegauge.py

```diff
@@ -0,0 +1,1279 @@
+        -> basic_stats(): Calculates multiple of the above metrics.
+
+        *Analysis*
+        -> resample_mean(): For resampling data in time using averaging
+        -> apply_doodson_xo_filter(): Remove tidal signal using Doodson XO
+        -> find_high_and_low_water(): Find maxima and minima of time series
+    """
+
+    def __init__(self, file_path: str = None, date_start=None, date_end=None, config: Union[Path, str] = None):
+        """
+        Initialise TIDEGAUGE object either as empty (no arguments) or by
+        reading GESLA data from a directory between two datetime objects.
+
+        Example usage:
+        --------------
+        # Read tide gauge data for data in January 1990
+        date0 = datetime.datetime(1990,1,1)
+        date1 = datetime.datetime(1990,2,1)
+        tg = coast.TIDEGAUGE(<'path_to_file'>, date0, date1)
+
+        # Access the data
+        tg.dataset
+
+        Parameters
+        ----------
+        file_path (list of str) : Filename to read from directory.
+        date_start (datetime) : Start date for data read. Optional
+        date_end (datetime) : end date for data read. Optional
+        config (Path or str) : configuration file
+
+        Returns
+        -------
+        Self
+        """
+        debug(f"Creating a new ..... {get_slug(self)}")
+        super().__init__(config)
+
+        # If file list is supplied, read files from directory
+        if file_path is None:
+            self.dataset = None
+        else:
+            self.dataset = self.read_gesla_to_xarray_v3(file_path,
+                                                        date_start, date_end)
+            self.apply_config_mappings()
+
+        print(f"{get_slug(self)} initialised")
+
+    ############ tide gauge methods ##############################################
+    @classmethod
+    def read_gesla_to_xarray_v3(cls, fn_gesla, date_start=None, date_end=None):
+        """
+        For reading from a single GESLA2 (Format version 3.0) file into an
+        xarray dataset. Formatting according to Woodworth et al. (2017).
+        Website: https://www.gesla.org/
+        If no data lies between the specified dates, a dataset is still created
+        containing information on the tide gauge, but the time dimension will
+        be empty.
+        Parameters
+        ----------
+        fn_gesla (str) : path to gesla tide gauge file
+        date_start (datetime) : start date for returning data
+        date_end (datetime) : end date for returning data
+
+        Returns
+        -------
+        xarray.Dataset object.
+        """
+        debug(f"Reading \"{fn_gesla}\" as a GESLA file with {get_slug(cls)}")  # TODO Maybe include start/end dates
+        try:
+            header_dict = cls.read_gesla_header_v3(fn_gesla)
+            dataset = cls.read_gesla_data_v3(fn_gesla, date_start, date_end)
+        except:
+            raise Exception('Problem reading GESLA file: ' + fn_gesla)
```

Comment:
Should we log the actual exception message here too? currently it isn't logged, and is masked by the custom exception.

Reply:
This is original code not dealt with. Will leave for later sprint to manage exceptions and file logging.